luna_soc/gateware/core/usb2/ulpi.py

```diff
@@ -0,0 +1,57 @@
+#
+# This file is part of LUNA.
+#
+# Copyright (c) 2020-2025 Great Scott Gadgets <info@greatscottgadgets.com>
+# SPDX-License-Identifier: BSD-3-Clause
+
+""" ULPI interfacing hardware. """
+
+from amaranth              import *
+from amaranth.lib          import wiring
+from amaranth.lib.wiring   import In, Out
+
+class Signature(wiring.Signature):
+    """ULPI bus signature.
+
+    Interface attributes
+    --------------------
+    i : :class:`Signal`
+        Input value of the ULPI data lines.
+    o : :class:`Signal`
+        Output value of the ULPI data lines.
+    oe : :class:`Signal`
+        True iff we're trying to drive the ULPI data lines.
+    clk : :class:`Signal`
+        The ULPI clock signal.
+    nxt : :class:`Signal`
+        The ULPI 'next' throttle signal.
+    stp : :class:`Signal`
+        The ULPI 'stop' event signal.
+    dir : :class:`Signal`
+        The ULPI bus-direction signal.
+    rst : :class:`Signal`
+        The ULPI 'reset' signal.
+    """
+
+    def __init__(self):
+        super().__init__({
+            "data" : In(
```

Comment:
This is incorrect; when you have an `In` signature member, the direction of its child members are flipped, so `data.i` ends up being an output, and vice versa.

Reply:
Argh, that hurts my head… but thank you!

Reply:
Fixed in: 84c942190ace95a8e2e4b8482cd2983fec014232